### PR TITLE
Student finance calculator/re 2017 18 updates

### PR DESCRIPTION
**Pull request by ikennaokpala**
Supersedes #2784

[Trello card](https://trello.com/c/B3lQ9yHW/361-student-finance-calculator-2017-18)

## Description
The aim of this PR is to make changes to the student finance calculator to include 2017-18 figures and remove 2015-16 figures.
The changes are:

- remove 2015-16 entirely
- add 2017-18 and figures for all student finance products
- add several new 2017-18 course options, and create new outcomes for most courses (but the content for these outcomes will mostly come from existing outcomes)
- Content changes to question and outcomes

## Factcheck

[Preview link](//smart-answers-pr-2798.herokuapp.com/student-finance-calculator)
[GOV.UK](//gov.uk/student-finance-calculator)

## Expected changes

* Removal of 2015-2016 legacy option
* Addition of 2017-2018 option
* Addition of 2017-2018 rates

### Before 

screen shot 2016-10-19 at 13 27 48

### After 

screen shot 2016-10-19 at 13 27 35

### CODE REVIEW

File: test/unit/calculators/student_finance_calculator_test.rb [hunk 1]
```diff
@@ -39,6 +39,33 @@ class StudentFinanceCalculatorTest < ActiveSupport::TestCase
...
       end
 
+      context "childcare_grant" do
+        context "for one child" do
+          context "in 2016-2017" do
+            should "be 155.24" do
+              calculator = StudentFinanceCalculator.new(
+                course_start: "2016-2017",
+                household_income: 25_000,
+                residence: :unused_variable
```
Comment (leenagupte) [inline test/unit/calculators/student_finance_calculator_test.rb]: As you are passing in a params hash to StudentFinanceCalculator that you don't need to pass in a dummy value for `residence`?

File: lib/smart_answer_flows/student-finance-calculator.rb [hunk 2]
```diff
@@ -6,12 +6,19 @@ def define
...
       satisfies_need "100133"
 
+      calculator = Calculators::StudentFinanceCalculator.new
+
       #Q1
       multiple_choice :when_does_your_course_start? do
         option :"2016-2017"
 
+        calculate :calculator do
+          calculator
```
Comment (leenagupte) [inline lib/smart_answer_flows/student-finance-calculator.rb]: I think you can move line 9 here.

e.g.

```
calculate :calculator do
  calculator = Calculators::StudentFinanceCalculator.new
end
```

File: lib/smart_answer_flows/student-finance-calculator.rb [hunk 3]
```diff
@@ -6,12 +6,19 @@ def define
...
+        end
+
         save_input_as :start_date
```
Comment (leenagupte) [inline lib/smart_answer_flows/student-finance-calculator.rb]: I don't think you need this anymore.

File: lib/smart_answer_flows/student-finance-calculator.rb [hunk 4]
```diff
@@ -82,7 +82,8 @@ def define
           calculator.maintenance_loan_amount
         end
```
Comment (leenagupte) [inline lib/smart_answer_flows/student-finance-calculator.rb]: I don't think you need calculate blocks for `maintenance_grant_amount` and `maintenance_loan_amount` anymore. You should be able to access the calculator directly where ever those amounts are needed.

File: test/unit/calculators/student_finance_calculator_test.rb (hunk 1 above)
Comment (ikennaokpala) [inline test/unit/calculators/student_finance_calculator_test.rb]: This was pre the refactor of the StudentFinanceCalculator.. I moved the commit order around to help understand the intent here..

File: lib/smart_answer_flows/student-finance-calculator.rb (hunk 3 above)
Comment (ikennaokpala) [inline lib/smart_answer_flows/student-finance-calculator.rb]: It is required.. still used in some outcomes

File: lib/smart_answer_flows/student-finance-calculator.rb (hunk 4 above)
Comment (ikennaokpala) [inline lib/smart_answer_flows/student-finance-calculator.rb]: Removing them would be ideal..

File: lib/smart_answer_flows/student-finance-calculator.rb (hunk 3 above)
Comment (leenagupte) [inline lib/smart_answer_flows/student-finance-calculator.rb]: Yes, I saw it used in a later commit. Thanks.

**Comment by leenagupte** [comment]
@ikennaokpala This is much better! So much easier to follow and review. Thanks! 👍 

I made a couple of inline comments. These are mainly cosmetic. e.g. The params passed into the tests etc.

I'm happy to sign this off with the changes as they are.

File: lib/smart_answer_flows/student-finance-calculator.rb (hunk 2 above)
Comment (ikennaokpala) [inline lib/smart_answer_flows/student-finance-calculator.rb]: Yeah, I understand your suggestion.. The scope for `calculator` will break across the flow

Comment (leenagupte) [inline lib/smart_answer_flows/student-finance-calculator.rb]: Break? It should be available in all blocks inside question methods and be accessible from the outcome templates.

File: test/unit/calculators/student_finance_calculator_test.rb (hunk 1 above)
Comment (leenagupte) [inline test/unit/calculators/student_finance_calculator_test.rb]: The commit order is good.

Comment (leenagupte) [inline test/unit/calculators/student_finance_calculator_test.rb]: It's just that, for this particular test, you only need to pass (or set) the `course_start`. The other two values are not needed.

**Pull request merged**